### PR TITLE
Add C code for Erfcx 

### PR DESCRIPTION
**Pull request by ricardoV94**
Mentioned in #267 

Here are a few important guidelines and requirements to check before your PR can be merged:
+ [x] There is an informative high-level description of the changes.
+ [x] The description and/or commit message(s) references the relevant GitHub issue(s).
+ [x] [`pre-commit`](https://pre-commit.com/#installation) is installed and [set up](https://pre-commit.com/#3-install-the-git-hook-scripts).
+ [x] The commit messages follow [these guidelines](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html).
+ [x] The commits correspond to [_relevant logical changes_](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes), and there are **no commits that fix changes introduced by other commits in the same branch/BR**.  If your commit description starts with "Fix...", then you're probably making this mistake.
+ [x] There are tests covering the changes introduced in the PR.

Don't worry, your PR doesn't need to be in perfect order to submit it.  As development progresses and/or reviewers request changes, you can always [rewrite the history](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History#_rewriting_history) of your feature/PR branches.

If your PR is an ongoing effort and you would like to involve us in the process, simply make it a [draft PR](https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/about-pull-requests#draft-pull-requests).

### CODE REVIEW

**Comment by ricardoV94** [comment]
I have tried without success to use directly the functions from the Fadddeeva library: http://ab-initio.mit.edu/wiki/index.php/Faddeeva_Package

Faddeeva.cc contains the code for the erfcx function, and it requires either Faddeeva.h (c compiler) or Faddeeva.hh (c++ compiler). When I try to compile the following function I get an error saying that Faddeeva.hh cannot be found (I am suprised aesara is compiling to c++ and not to c, which means I might have to change the function call...):

```python
x = aet.dscalar('x'); y = aet.erfcx(x); func = ae.function([x], y)
```
```
Traceback (most recent call last):
  File "/usr/lib/python3.8/code.py", line 90, in runcode
    exec(code, self.locals)
  File "<input>", line 1, in <module>
  File "/home/ricardo/Documents/Projects/Theano-PyMC/aesara/compile/function/__init__.py", line 337, in function
    fn = pfunc(
  File "/home/ricardo/Documents/Projects/Theano-PyMC/aesara/compile/function/pfunc.py", line 524, in pfunc
    return orig_function(
  File "/home/ricardo/Documents/Projects/Theano-PyMC/aesara/compile/function/types.py", line 1981, in orig_function
    fn = m.create(defaults)
  File "/home/ricardo/Documents/Projects/Theano-PyMC/aesara/compile/function/types.py", line 1836, in create
    _fn, _i, _o = self.linker.make_thunk(
  File "/home/ricardo/Documents/Projects/Theano-PyMC/aesara/link/basic.py", line 294, in make_thunk
```

I am able to compile locally to both c and c++:
```C
#include <stdio.h>
#include "Faddeeva.cc"

int main(){
        double result;

        result = Faddeeva_erfcx_re(0.99);
        printf("result = %f", result);
        return(0);
}
```
```bash
gcc -o examplec example.c -lm
./examplec
result = 0.430331
```
```C++
#include <stdio.h>
#include "Faddeeva.cc"

int main(){
        double result;

        result = Faddeeva::erfcx(0.99);
        printf("result = %f", result);
        return(0);
}
```
```bash
g++ -o examplecc example.cc
./examplecc
result = 0.430331
```

**Comment by twiecki** [comment]
How do you know it's a problem with the header? The exception is rather uninformative.

File: aesara/scalar/basic_scipy.py [hunk 1]
```diff
@@ -150,6 +150,34 @@ def L_op(self, inputs, outputs, grads):
         )
         return (gz * (-cst + (2.0 * x) * erfcx(x)),)
 
+    def c_header_dirs(self, **kwargs):
+        # Faddeeva.cc requires Faddeeva.hh (c++) or Faddeeva.h (c), which are in this folder
+        # From the error messages it seems aesara is compiling with c++
+        dirs = super().c_header_dirs(**kwargs) + [
+            os.path.join(os.path.dirname(__file__), "c_code")
+        ]
+        raise SystemError
```
Comment (twiecki) [inline aesara/scalar/basic_scipy.py:159]: Does this get triggered?

**Comment by ricardoV94** [comment]
> How do you know it's a problem with the header? The exception is rather uninformative.

Because I didn't copy the exception well :b. Let me update it

File: aesara/scalar/basic_scipy.py (hunk 1 above)
Comment (ricardoV94) [reply aesara/scalar/basic_scipy.py:159]: Nope

**Comment by brandonwillard** [comment]
If you're not already, you might need to clear the compilation cache&mdash;using the `aesara-cache` tool&mdash;between certain types of changes (or simply all of them, to be safe).

**Comment by ricardoV94** [comment]
> If you're not already, you might need to clear the compilation cache—using the `aesara-cache` tool—between certain types of changes (or simply all of them, to be safe).

I did clear it. I mean I think I did, because while I can do it in an environment where aesara in installed through pip, I cannot do it in the development environment, so I manually deleted the .aesara folder in the home directory.

**Comment by ricardoV94** [comment]
Am I misunderstanding the role of the `c_header_dirs method`? It never seems to be called. I tried adding it and raising a SystemError in other methods implemented in basic_scipy.py and it seems it's never triggered.

**Comment by dfm** [comment]
I can confirm that the header directory method is not called. I assume that it has something to do with this recently added condition: https://github.com/pymc-devs/aesara/blob/22c8c46dcb9898e5ca40b248692211b350568af6/aesara/link/c/basic.py#L1061

You can just replace

```c++
#  include "Faddeeva.hh"
```

in `Faddeeva.cc` with the some of contents of `Faddeeva.hh` to get it to work.

<details>
<summary>A subset of Faddeeva.hh</summary>

```c++
#include <complex>

namespace Faddeeva {

// compute w(z) = exp(-z^2) erfc(-iz) [ Faddeeva / scaled complex error func ]
extern std::complex<double> w(std::complex<double> z,double relerr=0);
extern double w_im(double x); // special-case code for Im[w(x)] of real x

// Various functions that we can compute with the help of w(z)

// compute erfcx(z) = exp(z^2) erfc(z)
extern std::complex<double> erfcx(std::complex<double> z, double relerr=0);
extern double erfcx(double x); // special case for real x

// compute erf(z), the error function of complex arguments
extern std::complex<double> erf(std::complex<double> z, double relerr=0);
extern double erf(double x); // special case for real x

// compute erfi(z) = -i erf(iz), the imaginary error function
extern std::complex<double> erfi(std::complex<double> z, double relerr=0);
extern double erfi(double x); // special case for real x

// compute erfc(z) = 1 - erf(z), the complementary error function
extern std::complex<double> erfc(std::complex<double> z, double relerr=0);
extern double erfc(double x); // special case for real x

// compute Dawson(z) = sqrt(pi)/2  *  exp(-z^2) * erfi(z)
extern std::complex<double> Dawson(std::complex<double> z, double relerr=0);
extern double Dawson(double x); // special case for real x

} // namespace Faddeeva
```
</details>

You probably also want the `c_code` method to be something like:

```python
    def c_code(self, node, name, inp, out, sub):
        (x,) = inp
        (z,) = out
        if node.inputs[0].type in float_types:
            dtype = "npy_" + node.outputs[0].dtype
            return (
                """%(z)s =
                (%(dtype)s) Faddeeva::erfcx(%(x)s);"""
                % locals()
            )
        raise NotImplementedError("only floatingpoint is implemented")
```

**Comment by ricardoV94** [comment]
@dfm that is really helpful! Thanks for digging around. 

Since you were the one who first mentioned this library, do you think there is any advantage in using the Faddeeva version vs the Scipy adaptation. I did a quick comparison recently and the scipy guys seem to have replace some constructs like npy_nan / npy_inf as well as to reduce the number of terms used in some of the approximations: https://github.com/scipy/scipy/blob/master/scipy/special/Faddeeva.cc

On the other hand the Faddeeva version seems to have been updated slightly since then.

**Comment by twiecki** [comment]
Seems like a bug then, no? Or you need to inherit from `CLinkerObject` somehow.

**Comment by dfm** [comment]
@ricardoV94: I'm not sure that I have any insight on that - sorry!

@twiecki: perhaps. I would have expected that conditional to work because I think it does inherit appropriately (although there is some multiple inheritance and my intuition is often wrong!) somewhere in the hierarchy, but that's the only spot I could find where this might be coming from. It probably deserves some experiments/a test.

**Comment by ricardoV94** [comment]
I will check if that's where it fails, but in terms of inheritance things look about right:

```python
isinstance(aet.erfcx, aesara.link.c.interface.CLinkerObject)  #True
```

**Comment by ricardoV94** [comment]
Interestingly enough the c_header_dirs is reached when using a scalar:

```python
import aesara as ae; import aesara.tensor as aet; import aesara.scalar as aes
x = aet.dscalar('x'); y = aes.erfcx(x); func = ae.function([x], y)
```

**Comment by ricardoV94** [comment]
So `Elemwise` was not calling the `c_header_dirs`.

This change seems to fix things: https://github.com/pymc-devs/aesara/pull/317/commits/cdb8010f32c9b53de47996d5ca0d0e4a6284c329

Is there a reason why `c_header_dirs` shoud not be called for `Elemwise`?

**Comment by twiecki** [comment]
I would put a break point into https://github.com/pymc-devs/aesara/blob/22c8c46dcb9898e5ca40b248692211b350568af6/aesara/link/c/basic.py#L1060 and check what's happening in both cases (scalar and vector).

**Comment by twiecki** [comment]
Great catch, I'm sure that's just an oversight.

**Comment by twiecki** [comment]
Then I think we just need some tests 🥳

**Comment by ricardoV94** [comment]
Actually the old tests should already be running with the C code version now (they failed when the c_code was invalid).

We need to decide on whether to use the Faddeeva or the Scipy implementation, which makes use of numpy nan / inf / copysign constructs instead (and uses less approximation terms in some places).

**Comment by twiecki** [comment]
So instead of this file you'd copy over the version modified by scipy? It sounds like a better implementation so should probably go with twhat. what do you think?

**Comment by ricardoV94** [comment]
> So instead of this file you'd copy over the version modified by scipy? It sounds like a better implementation so should probably go with twhat. what do you think?

I will check again the diff. Maybe I will merge the more recent changes that increase accuracy / optimizations on the Faddeeva implementation with the numpy macros for nan/inf/copysign on the scipy implementation.

**Comment by ricardoV94** [comment]
Something still funny with the header. 

`py.test tests/tensor/test_basic_scipy.py -k "TestErfcxInplaceBroadcast"` passes
`py.test tests/tensor/test_basic_scipy.py -k "TestErfcxBroadcast" fails with the same "Faddeeva.hh" missing story.

Some other place where the call to c_header_dirs() is missing?

**Comment by brandonwillard** [comment]
> We need to decide on whether to use the Faddeeva or the Scipy implementation, which makes use of numpy nan / inf / copysign constructs instead, and uses less approximation terms in some places.

In general, we do not want to add new external dependencies, so the SciPy approach may be preferable.

**Comment by ricardoV94** [comment]
Any tips on where to look to understand why the inplace tests pass but not the no-inplace?

Sounds like the same header for missing thing, but I have no idea where to look.

**Comment by brandonwillard** [comment]
The problem might be `aesara.scalar.Composite`.

Individual scalar `Op`s are combined into single `Composite` `Op`s, and that doesn't implement the `c_header_dirs` method.

**Comment by brandonwillard** [comment]
OK, that seems to have fixed it.  

I cleaned up the commits, so when/if these test pass and we're sure that we want to include this Faddeeva code in Aesara (e.g. we need to make sure that the license works out, that the existing SciPy dependency isn't just as good, etc.), we should be good to merge.

**Comment by ricardoV94** [comment]
> OK, that seems to have fixed it.  
> 
> I cleaned up the commits, so when/if these test pass and we're sure that we want to include this Faddeeva code in Aesara (e.g. we need to make sure that the license works out, that the existing SciPy dependency isn't just as good, etc.), we should be good to merge.

Thanks a lot for your help. I will double check the license tomorrow and compare the scipy and faddeeva versions. We should be ready to go soon

**Comment by twiecki** [comment]
Excited to get this speed-up in!

File: aesara/scalar/c_code/Faddeeva.hh [hunk 2]
```diff
@@ -0,0 +1,62 @@
+/* Copyright (c) 2012 Massachusetts Institute of Technology
```
Comment (twiecki) [inline aesara/scalar/c_code/Faddeeva.hh:1]: Do we need a `.h` *and* a `.hh` file?

Comment (ricardoV94) [reply aesara/scalar/c_code/Faddeeva.hh:1]: No, only the .hh

**Comment by ricardoV94** [comment]
I ended up using the Faddeeva original files.

The scipy version was giving me problems with the npy_copysign, and it seems they already had to revert the npy_isinf and npy_isnan functions in the past due to conflicts with the `cmath` library (see https://github.com/scipy/scipy/issues/5689). Overall it seems that most of the benefits of using the numpy implementations are limited due to conflicts with the c++ standard libraries. 

One advantage of the Faddeeva version is that some small precision and performance improvements have been done since the scipy branch was last updated.

Also, these functions are implemented with complex numbers, so if there is a demand to have them work with aesara complex objects it should be possible to do it. I am happy to NOT do it, unless you guys think this should be done now.

License is MIT so it should be fine to use: http://ab-initio.mit.edu/wiki/index.php/Faddeeva_Package#License

Finally, just for sanity check I ran the built-in tests in the source file and everything passed.

**Comment by twiecki** [comment]
Sounds like the right move. We don't need complex support.

**Review by twiecki — APPROVED** [review]

**Comment by ricardoV94** [comment]
Should we add a cache number?

**Comment by twiecki** [comment]
What's that?

**Comment by ricardoV94** [comment]
https://theano-pymc.readthedocs.io/en/latest/extending/extending_theano_c.html?highlight=c%20cache#c_code_cache_version

**Comment by twiecki** [comment]
Sure, good idea.

**Review by brandonwillard — CHANGES_REQUESTED** [review]
If we're going to choose between header files, we should go with the C one (i.e. `Faddeeva.h`), since this is the approach that the rest of the library takes.

**Comment by brandonwillard** [comment]
Also, this PR seems to have at least doubled in commits.  We only need three: i.e. one corresponding to the Faddeeva/erfcx addition and one for each of the newly added header methods.

**Comment by ricardoV94** [comment]
> Also, this PR seems to have at least doubled in commits.  We only need three: i.e. one corresponding to the Faddeeva/erfcx addition and one for each of the newly added header methods.

I didn't pay attention to this. Just assumed squash and merge is going on. I'll clean them once we decide about the header.

**Comment by twiecki** [comment]
Thanks @ricardoV94, this is a great contribution!

**Comment by brandonwillard** [comment]
These commits needed to be squashed; there were some odd duplicates in there.

**Pull request merged**